### PR TITLE
[FEATURE] Implement addon black- and whitelist

### PR DESCRIPTION
**Pull request by dschmidt**
... per RFC https://github.com/ember-cli/rfcs/blob/master/active/0000-addon-black-and-whitelist-for-apps.md

### CODE REVIEW

File: lib/broccoli/ember-app.js [hunk 1]
```diff
@@ -341,7 +341,15 @@ EmberApp.prototype._notifyAddonIncluded = function() {
   this.project.addons = this.project.addons.filter(function(addon) {
     addon.app = this;
 
-    if (!addon.isEnabled || addon.isEnabled()) {
+    var addonEnabled = !addon.isEnabled || addon.isEnabled();
+
+    if (addonEnabled && this.options.addonBlacklist && this.options.addonBlacklist.indexOf(addon.name) !== -1) {
```
Comment (stefanpenner) [inline lib/broccoli/ember-app.js]: lots of complexity in these lines, let me recommend this be extract to its own function, which then itself can be nicely unit tested.

**Comment by stefanpenner** [comment]
question, thoughts on:

```js
addons: {
  blacklist: [...],
  whiltelist: [...]
};
```
?

File: lib/broccoli/ember-app.js (hunk 1 above)
Comment (dschmidt) [inline lib/broccoli/ember-app.js]: Okies, will do.

**Comment by dschmidt** [comment]
Dunno, I don't care tbh :trollface:

**Comment by stefanpenner** [comment]
i prefer `addons: { ...}` @rwjblue r?

**Comment by dschmidt** [comment]
The naming `disabledByBlacklisted` and `disabledByWhitelist` might seem weird at first, but even while writing the code `blacklisted` and `whitelisted` seemed very ambiguous.
Blacklisted would be rather clear, but what is whitelisted? Is an addon only whitelisted if there is a whitelist? If so, ifs get rather complicated and unreadable again.

This is why I went for these rather bulky var names.

**Comment by stefanpenner** [comment]
Should we also error if an addon is mentioned, that is not actually an addon? Might be a great way to catch bugs.

**Comment by dschmidt** [comment]
I'm afraid, you are right. I thought about it, but was too lazy ... will do :P

**Comment by dschmidt** [comment]
Updated

**Comment by stefanpenner** [comment]
SGTM

**Comment by dschmidt** [comment]
Cool, will squash commits once tests are complete

**Comment by rwjblue** [comment]
LGTM

**Comment by stefanpenner** [comment]
@homu r+

**Pull request merged**